src/UsgsAstroSarSensorModel.cpp

```diff
@@ -136,3 +141,209 @@ string UsgsAstroSarSensorModel::constructStateFromIsd(
+                  double tolerance, double maxIters);
+
+csm::ImageCoord UsgsAstroSarSensorModel::groundToImage(
+    const csm::EcefCoord& groundPt,
+    double desiredPrecision,
+    double* achievedPrecision,
+    csm::WarningList* warnings) const {
+
+  //MESSAGE_LOG(this->m_logger, "Computing groundToImage(ImageCoord) for {}, {}, {}, with desired precision {}",
+  //          groundPt.x, groundPt.y, groundPt.z, desiredPrecision);
+
+  // Find time of closest approach to groundPt and the corresponding slant range by finding 
+  // the root of the doppler shift frequency
+  csm::EcefCoord newGroundPt(groundPt);
+  double time = dopplerShiftRoot(newGroundPt);
+  double slantRange = calcSlantRange(newGroundPt, time);
+
+  if (time != -1) {
+    // Find the ground range, based on the ground-range-to-slant-range polynomial defined by the 
+    // range coefficient set, with a time closest to the calculated time of closest approach
+    double groundRange = slantToGroundRange(groundPt, time, slantRange);
+
+    if (groundRange =! -1) {
+      double line = 1 + (time - START_TIME) / EXPOSURE_DURATION;
+      double sample = 1 + groundRange / SCALED_PIXEL_WIDTH;
+      return csm::ImageCoord(line, sample);
+    }
+  }
+  return csm::ImageCoord(-1, -1); // what do we do if this fails?
+}
+
+// Calculate the root 
+double UsgsAstroSarSensorModel::dopplerShiftRoot(
+    csm::EcefCoord groundPt) const{
+
+  // Moon body-fixed coordinates of surface point
+  csm::EcefCoord surfPt(0.0, 0.0, 0.0); // obviously populate with real value
+
+   std::function<double(double)> dopplerShiftFrequency = [surfPt](double time) { 
+     // Exsiting vector subtraction function in usgscsm? 
+     double spacecraftPosition[3] = {3.73740000e+06,  0.00000000e+00, -0.00000000e+00}; //spacecraftPosition(time);
+     double spacecraftVelocity[3] = {-0.00000000e+00,  0.00000000e+00, -3.73740000e+0};; //spacecraftPosition(time);
+     double lookVector[3];
+
+     lookVector[0] = spacecraftPosition[0] - surfPt.x;
+     lookVector[1] = spacecraftPosition[1] - surfPt.y;
+     lookVector[2] = spacecraftPosition[2] - surfPt.z;
+
+     // Existing vector magnitude function in usgscsm? 
+     double slantRange = sqrt(pow(lookVector[0], 2) +  pow(lookVector[1], 2) + pow(lookVector[2], 2)); 
+
+
+     // Existing dot product function in usgscsm?
+     double dopplerShift = -2.0 * (lookVector[0]*spacecraftVelocity[0] + lookVector[1]*spacecraftVelocity[1] 
+                            + lookVector[2]*spacecraftVelocity[2])/(slantRange * WAVELENGTH);
+     return dopplerShift;
+   };
+
+  // Lower-bound for doppler-shift
+  double startShift = dopplerShiftFrequency(START_TIME);
+
+  // Upper-bound for doppler-shift
+  double stopShift = dopplerShiftFrequency(STOP_TIME);
+
+  // Make sure we bound root (dopplerShift = 0.0)
+  if (((startShift < 0.0) && (stopShift < 0.0)) || ((startShift > 0.0) && (stopShift > 0.0))) {
+    return -1;
```

Comment:
Instead of doing C-style return code, this should just raise an exception that gets caught higher up.